src/datasets/formatting/formatting.py

```diff
@@ -260,6 +262,34 @@ def decode_batch(self, batch: pd.DataFrame) -> pd.DataFrame:
         return self.decode_row(batch)
 
 
+class LazyDict(UserDict):
+    def __init__(self, data: dict, formatter: "Formatter"):
+        self.data = data
```

Comment:
Here data is the "extracted" data - e.g. a python dictionary from an arrow table.

We could also imagine storing the arrow data instead, and lazily apply the full formatting (extraction + decoding + type formatting)

Maybe it's not the scope of this PR, but it would make all the map/filter functions ultra fast if you only need to read one column.

Right now you have to handle arrow data, extracted data, decoded data and formatted data.
Maybe we only need to handle arrow data and formatted data if it makes things simpler

Reply:
Had the same idea. I think I'll explore that as part of this PR.